rfc/rfc-78/rfc-78.md

```diff
@@ -0,0 +1,301 @@
+- If there are new MDT partitions (functional index, sec index), nuke them. Update table properties.
+- Check latest snapshot files for log files w/ new features enabled. For eg, deletion vector, partial update. If any log files are found in latest snapshot, we have to trigger compaction for those file groups. This calls for a custom compaction strategy as well which compacts only for certain file groups.
+   - In order to reduce downtime to downgrade, we are inclining to do this only for snapshot reads. If there are older file slices and users are issuing time travel or incremental reads, and if they have any of the new features enabled (deletion vector, partial update), readers could break after downgrade.
+
+## Implementation details
+
+Atleast for adding read capability of 1.x tables in 0.16.0, documenting the changes required at class level. Other two sections (0.16.0 upgrade to 1.x or 1.x downgrade to 0.16.0) should be fairly straightforward.
+
+### Timeline changes:
+Let’s reiterate what we need to support w/ 0.16.0 reader.
+
+### Timeline read of 1.x need to be supported
+- Commit instants w/ completion time should be readable. HoodieInstant parsing logic to parse completion time should be ported over.
+- We could ignore the completion time log file ordering semantics since we don’t plan to support it yet in 0.16.0. But reading should not break.
+- Pending Clustering commits using “cluster” action should be readable in 0.16.0 reader.
+- HoodieDefaultTimeline should be able to support both 0.x timeline and 1.x timeline. 
+- Commit metadata in both avro and json should be supported. The challenging part here is, for every commit metadata, we might have to deserialize to avro and on exception try json. We could deduce the format using completion file name, but as per current code layering, deserialization methods does not know the file name( method takes byte[]). Similarly for clustering commits, unless we have some kind of watermark, we have to keep considering replace commits as well in the FSV building logic to ensure we do not miss any pending clustering commits.
+
+Actual code changes: 
+- So for above mentioned reasons, we plan to introduce a new table config named “table version upgrade commit time” which will be referring to the latest commit which was written in 0.16.0 just before upgrading to 1.x. Once we have this commit time as water mark, which marks the upgrade in the timeline, even a 1.x reader should only account for older compatible reads up until the water mark. For eg, to deser pending replace commits for the interest of clustering can be dropped after the water mark. Once we get past that, only “cluster”ing pending commits can represent pending clustering writes. We could use the same watermark for commit metadata parsing as well. So, that at some point(after the water mark) we could safely assume entire timeline is only in avro (speaking from a 1.x FSV/reader standpoint) and not in json.
```

Comment:
update: we are thinking to avoid the using it for commit metadata deser. 
a 0.16.x reader will favor json and then fallback to avro. 
0.1.x reader will favor avro and then fallback json.